tests/view/domconverter/view-to-dom.js

```diff
@@ -164,10 +166,192 @@ describe( 'DomConverter', () => {
+				const viewDiv = new ViewContainerElement( 'div', null, [
+					new ViewContainerElement( 'p', null, new ViewText( ' foo' ) ),
+					new ViewContainerElement( 'p', null, new ViewText( 'bar' ) ),
+					new ViewContainerElement( 'p', null, new ViewText( ' xxx' ) )
+				] );
+
+				const domDiv = converter.viewToDom( viewDiv, document );
+
+				expect( domDiv.innerHTML ).to.equal( '<p>&nbsp;foo</p><p>bar</p><p>&nbsp;xxx</p>' );
+			} );
+
+			it( 'at the end of each container element', () => {
+				const viewDiv = new ViewContainerElement( 'div', null, [
+					new ViewContainerElement( 'p', null, new ViewText( 'foo ' ) ),
+					new ViewContainerElement( 'p', null, new ViewText( 'bar' ) ),
+					new ViewContainerElement( 'p', null, new ViewText( 'xxx ' ) )
+				] );
+
+				const domDiv = converter.viewToDom( viewDiv, document );
+
+				expect( domDiv.innerHTML ).to.equal( '<p>foo&nbsp;</p><p>bar</p><p>xxx&nbsp;</p>' );
+			} );
+
+			it( 'when there are multiple spaces next to each other or between attribute elements', () => {
+				const viewDiv = new ViewContainerElement( 'div', null, [
+					new ViewText( 'x  x   x x ' ),
+					new ViewAttributeElement( 'b', null, new ViewText( ' x ' ) ),
+					new ViewAttributeElement( 'i', null,
+						new ViewAttributeElement( 'b', null,
+							new ViewAttributeElement( 'u' , null, new ViewText( ' x' ) )
+						)
+					)
+				] );
+
+				const domDiv = converter.viewToDom( viewDiv, document );
+
+				expect( domDiv.innerHTML ).to.equal( 'x &nbsp;x &nbsp; x x <b>&nbsp;x </b><i><b><u>&nbsp;x</u></b></i>' );
+			} );
+
+			it( 'all together', () => {
+				const viewDiv = new ViewContainerElement( 'div', null, [
+					new ViewContainerElement( 'p', null, [
+						new ViewText( ' x  x   x x ' ),
+						new ViewAttributeElement( 'b', null, new ViewText( ' x ' ) ),
+						new ViewAttributeElement( 'i', null,
+							new ViewAttributeElement( 'b', null,
+								new ViewAttributeElement( 'u' , null, new ViewText( ' x ' ) )
+							)
+						)
+					] ),
+					new ViewContainerElement( 'p', null, new ViewText( '  x  ' ) )
+				] );
+
+				const domDiv = converter.viewToDom( viewDiv, document );
+
+				expect( domDiv.innerHTML ).to.equal(
+					'<p>&nbsp;x &nbsp;x &nbsp; x x <b>&nbsp;x </b><i><b><u>&nbsp;x&nbsp;</u></b></i></p><p>&nbsp; x &nbsp;</p>'
+				);
+			} );
+
+			it( 'spaces in a text node', () => {
+				function test( inputTexts, output ) {
+					const viewElement = new ViewContainerElement( 'p' );
+
+					if ( typeof inputTexts == 'string' ) {
+						inputTexts = [ inputTexts ];
+					}
+
+					for ( let text of inputTexts ) {
+						viewElement.appendChildren( new ViewText( text.replace( /_/g, '\u00A0' ) ) );
+					}
+
+					const domElement = converter.viewToDom( viewElement, document );
+					const data = domElement.innerHTML.replace( /&nbsp;/g, '_' );
+
+					expect( data ).to.equal( output );
+				}
+
+				// At the beginning.
+				test( ' x', '_x' );
```

Comment:
Those `test()` function should be moved out of `it()`. They should call their own `it()` with proper names. Otherwise you won't be able to know what's actually failing. And it's super frustraiting if e.g. you know that 3rd cases fails but you don't know which of the next ones fails too, because first failed one stops test execution.